File: backend/src/main/java/com/agentworkbench/permission/service/PermissionService.java
package com.agentworkbench.permission.service;

import com.agentworkbench.permission.entity.Permission;
import com.agentworkbench.permission.entity.Role;
import com.agentworkbench.permission.entity.RolePermission;
import com.agentworkbench.permission.entity.UserRole;
import com.agentworkbench.permission.mapper.PermissionMapper;
import com.agentworkbench.permission.mapper.RoleMapper;
import com.agentworkbench.permission.mapper.RolePermissionMapper;
import com.agentworkbench.permission.mapper.UserRoleMapper;
import com.agentworkbench.common.exception.BusinessException;
import com.agentworkbench.common.result.ErrorCode;
import com.agentworkbench.user.mapper.UserMapper;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

@Service
@RequiredArgsConstructor
public class PermissionService {

    private final RoleMapper roleMapper;
    private final PermissionMapper permissionMapper;
    private final RolePermissionMapper rolePermissionMapper;
    private final UserRoleMapper userRoleMapper;
    private final UserMapper userMapper;

    public List<Role> listRoles() {
        return roleMapper.selectList(null);
    }

    public Role getRole(Long id) {
        return roleMapper.selectById(id);
    }

    public Role createRole(String name, String code, String description) {
        Role role = new Role();
        role.setName(name);
        role.setCode(code);
        role.setDescription(description);
        roleMapper.insert(role);
        return role;
    }

    public Role updateRole(Long id, String name, String description) {
        Role role = roleMapper.selectById(id);
        if (role == null) return null;
        if (name != null) role.setName(name);
        if (description != null) role.setDescription(description);
        roleMapper.updateById(role);
        return role;
    }

    public List<Permission> listPermissions() {
        return permissionMapper.selectList(null);
    }

    public List<Long> getRolePermissionIds(Long roleId) {
        return rolePermissionMapper.selectList(
                        new QueryWrapper<RolePermission>().eq("role_id", roleId))
                .stream()
                .map(RolePermission::getPermissionId)
                .toList();
    }

    public Long countRoleUsers(Long roleId) {
        return userRoleMapper.selectCount(new QueryWrapper<UserRole>().eq("role_id", roleId));
    }

    @Transactional
    public void assignPermissions(Long roleId, List<Long> permissionIds) {
        rolePermissionMapper.delete(
                new QueryWrapper<RolePermission>().eq("role_id", roleId));
        for (Long permId : permissionIds) {
            RolePermission rp = new RolePermission();
            rp.setRoleId(roleId);
            rp.setPermissionId(permId);
            rolePermissionMapper.insert(rp);
        }
    }

    @Transactional
    public void assignRoles(Long userId, List<Long> roleIds) {
        if (roleIds == null || roleIds.isEmpty()) {
            throw new BusinessException(ErrorCode.PARAM_INVALID, "至少分配一个角色");
        }
        userRoleMapper.delete(
                new QueryWrapper<UserRole>().eq("user_id", userId));
        for (Long roleId : roleIds) {
            UserRole ur = new UserRole();
            ur.setUserId(userId);
            ur.setRoleId(roleId);
            userRoleMapper.insert(ur);
        }
    }

    public List<Long> getUserRoleIds(Long userId) {
        return userRoleMapper.selectList(
                        new QueryWrapper<UserRole>().eq("user_id", userId))
                .stream()
                .map(UserRole::getRoleId)
                .toList();
    }

    public List<Role> getUserRoles(Long userId) {
        List<Long> roleIds = getUserRoleIds(userId);
        if (roleIds.isEmpty()) {
            return List.of();
        }
        return roleMapper.selectBatchIds(roleIds);
    }

    public Map<Long, List<Role>> batchGetUserRoles(List<Long> userIds) {
        if (userIds == null || userIds.isEmpty()) {
            return Map.of();
        }
        List<UserRole> userRoles = userRoleMapper.selectList(
                new QueryWrapper<UserRole>().in("user_id", userIds));
        if (userRoles.isEmpty()) {
            return Map.of();
        }

        List<Long> roleIds = userRoles.stream().map(UserRole::getRoleId).distinct().toList();
        Map<Long, Role> roleMap = new HashMap<>();
        for (Role role : roleMapper.selectBatchIds(roleIds)) {
            roleMap.put(role.getId(), role);
        }

        Map<Long, List<Role>> result = new HashMap<>();
        for (UserRole userRole : userRoles) {
            Role role = roleMap.get(userRole.getRoleId());
            if (role != null) {
                result.computeIfAbsent(userRole.getUserId(), ignored -> new ArrayList<>()).add(role);
            }
        }
        return result;
    }

    public void assertCanDisableUser(Long targetUserId, Long currentUserId) {
        if (Objects.equals(targetUserId, currentUserId)) {
            throw new BusinessException(ErrorCode.CANNOT_DISABLE_SELF);
        }
        Role adminRole = getAdminRole();
        if (adminRole == null || !userHasRole(targetUserId, adminRole.getId())) {
            return;
        }
        if (countOtherActiveAdmins(adminRole.getId(), targetUserId) == 0) {
            throw new BusinessException(ErrorCode.CANNOT_REMOVE_LAST_ADMIN);
        }
    }

    public void assertCanChangeRoles(Long userId, List<Long> newRoleIds) {
        Role adminRole = getAdminRole();
        if (adminRole == null) {
            return;
        }
        boolean hadAdmin = userHasRole(userId, adminRole.getId());
        boolean willHaveAdmin = newRoleIds.contains(adminRole.getId());
        if (hadAdmin && !willHaveAdmin && countOtherActiveAdmins(adminRole.getId(), userId) == 0) {
            throw new BusinessException(ErrorCode.CANNOT_REMOVE_LAST_ADMIN);
        }
    }

    private Role getAdminRole() {
        return roleMapper.selectOne(new QueryWrapper<Role>().eq("code", "ADMIN"));
    }

    private boolean userHasRole(Long userId, Long roleId) {
        Long count = userRoleMapper.selectCount(
                new QueryWrapper<UserRole>()
                        .eq("user_id", userId)
                        .eq("role_id", roleId));
        return count > 0;
    }

    private long countOtherActiveAdmins(Long adminRoleId, Long excludeUserId) {
        List<UserRole> adminBindings = userRoleMapper.selectList(
                new QueryWrapper<UserRole>().eq("role_id", adminRoleId));
        return adminBindings.stream()
                .map(UserRole::getUserId)
                .filter(userId -> !Objects.equals(userId, excludeUserId))
                .map(userMapper::selectById)
                .filter(user -> user != null && user.getStatus() != null && user.getStatus() == 1)
                .count();
    }

    /**
     * Check if a user has a specific permission
     */
    public boolean hasPermission(Long userId, String permissionCode) {
        // Get user's roles
        List<UserRole> userRoles = userRoleMapper.selectList(
                new QueryWrapper<UserRole>().eq("user_id", userId));
        if (userRoles.isEmpty()) return false;

        List<Long> roleIds = userRoles.stream().map(UserRole::getRoleId).toList();

        // Get permissions for those roles
        List<RolePermission> rolePerms = rolePermissionMapper.selectList(
                new QueryWrapper<RolePermission>().in("role_id", roleIds));
        if (rolePerms.isEmpty()) return false;

        List<Long> permIds = rolePerms.stream().map(RolePermission::getPermissionId).toList();

        // Check if the permission code exists
        Long count = permissionMapper.selectCount(
                new QueryWrapper<Permission>()
                        .in("id", permIds)
                        .eq("code", permissionCode));
        return count > 0;
    }

    /**
     * Get all permission codes for a user
     */
    public List<String> getUserPermissionCodes(Long userId) {
        List<UserRole> userRoles = userRoleMapper.selectList(
                new QueryWrapper<UserRole>().eq("user_id", userId));
        if (userRoles.isEmpty()) return List.of();

        List<Long> roleIds = userRoles.stream().map(UserRole::getRoleId).toList();

        List<RolePermission> rolePerms = rolePermissionMapper.selectList(
                new QueryWrapper<RolePermission>().in("role_id", roleIds));
        if (rolePerms.isEmpty()) return List.of();

        List<Long> permIds = rolePerms.stream().map(RolePermission::getPermissionId).toList();

        List<Permission> permissions = permissionMapper.selectBatchIds(permIds);
        return permissions.stream().map(Permission::getCode).toList();
    }
}

File: backend-ts/src/permission/permission.service.ts
import { BusinessException } from '../common/business-exception.js';
import { ErrorCode } from '../common/error-code.js';
import type {
  Permission,
  PermissionRepository,
  Role,
  RolePermissionRepository,
  RoleRepository,
  UserRepository,
  UserRole,
  UserRoleRepository,
} from '../user/types.js';

export class PermissionService {
  constructor(
    private readonly roleRepo: RoleRepository,
    private readonly permissionRepo: PermissionRepository,
    private readonly rolePermissionRepo: RolePermissionRepository,
    private readonly userRoleRepo: UserRoleRepository,
    private readonly userRepo: UserRepository,
  ) {}

  listRoles(): Promise<Role[]> {
    return this.roleRepo.findAll();
  }

  getRole(id: number): Promise<Role | null> {
    return this.roleRepo.findById(id);
  }

  async createRole(name: string, code: string, description?: string | null): Promise<Role> {
    const role: Role = { name, code, description };
    const id = await this.roleRepo.insert(role);
    role.id = id;
    return role;
  }

  async updateRole(id: number, name?: string | null, description?: string | null): Promise<Role | null> {
    const role = await this.roleRepo.findById(id);
    if (!role) {
      return null;
    }
    if (name != null) {
      role.name = name;
    }
    if (description != null) {
      role.description = description;
    }
    await this.roleRepo.updateById(role);
    return role;
  }

  listPermissions(): Promise<Permission[]> {
    return this.permissionRepo.findAll();
  }

  async getRolePermissionIds(roleId: number): Promise<number[]> {
    const rows = await this.rolePermissionRepo.findByRoleId(roleId);
    return rows.map((r) => r.permissionId);
  }

  countRoleUsers(roleId: number): Promise<number> {
    return this.userRoleRepo.countByRoleId(roleId);
  }

  async assignPermissions(roleId: number, permissionIds: number[]): Promise<void> {
    // 先删后插必须同事务：中途失败不得留下「已删未插」的半截状态
    if (this.rolePermissionRepo.transaction) {
      await this.rolePermissionRepo.transaction(async (tx) => {
        await tx.deleteByRoleId(roleId);
        for (const permId of permissionIds) {
          await tx.insert({ roleId, permissionId: permId });
        }
      });
      return;
    }
    await this.rolePermissionRepo.deleteByRoleId(roleId);
    for (const permId of permissionIds) {
      await this.rolePermissionRepo.insert({ roleId, permissionId: permId });
    }
  }

  async assignRoles(userId: number, roleIds: number[] | null | undefined): Promise<void> {
    if (roleIds == null || roleIds.length === 0) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '至少分配一个角色');
    }
    // 先删后插必须同事务：中途失败不得留下「已删未插」的半截状态
    if (this.userRoleRepo.transaction) {
      await this.userRoleRepo.transaction(async (tx) => {
        await tx.deleteByUserId(userId);
        for (const roleId of roleIds) {
          await tx.insert({ userId, roleId });
        }
      });
      return;
    }
    await this.userRoleRepo.deleteByUserId(userId);
    for (const roleId of roleIds) {
      await this.userRoleRepo.insert({ userId, roleId });
    }
  }

  /**
   * 最后管理员保护 + 角色写入同一事务。
   * 对 ADMIN 绑定行 FOR UPDATE，避免并发双降级同时通过 count 检查后清空全部管理员。
   */
  async changeRolesWithAdminGuard(userId: number, newRoleIds: number[]): Promise<void> {
    if (newRoleIds.length === 0) {
      throw new BusinessException(ErrorCode.PARAM_INVALID, '至少分配一个角色');
    }
    if (!this.userRoleRepo.transaction || !this.userRoleRepo.findByRoleIdForUpdate) {
      // 无事务能力时退化为「先断言再写」（与旧行为一致）
      await this.assertCanChangeRoles(userId, newRoleIds);
      await this.assignRoles(userId, newRoleIds);
      return;
    }
    await this.userRoleRepo.transaction(async (tx) => {
      const adminRole = await this.getAdminRole();
      if (adminRole && tx.findByRoleIdForUpdate) {
        // 锁定 ADMIN 全部绑定行：并发降级在此串行化
        const bindings = await tx.findByRoleIdForUpdate(adminRole.id!);
        const hadAdmin = bindings.some((b) => b.userId === userId);
        const willHaveAdmin = newRoleIds.includes(adminRole.id!);
        if (hadAdmin && !willHaveAdmin) {
          let otherActive = 0;
          for (const b of bindings) {
            if (b.userId === userId) continue;
            const u = await this.userRepo.findById(b.userId);
            if (u && u.status === 1) otherActive += 1;
          }
          if (otherActive === 0) {
            throw new BusinessException(ErrorCode.CANNOT_REMOVE_LAST_ADMIN);
          }
        }
      }
      await tx.deleteByUserId(userId);
      for (const roleId of newRoleIds) {
        await tx.insert({ userId, roleId });
      }
    });
  }

  async getUserRoleIds(userId: number): Promise<number[]> {
    const rows = await this.userRoleRepo.findByUserId(userId);
    return rows.map((r) => r.roleId);
  }

  async getUserRoles(userId: number): Promise<Role[]> {
    const roleIds = await this.getUserRoleIds(userId);
    if (roleIds.length === 0) {
      return [];
    }
    return this.roleRepo.findByIds(roleIds);
  }

  async batchGetUserRoles(userIds: number[] | null | undefined): Promise<Map<number, Role[]>> {
    const result = new Map<number, Role[]>();
    if (userIds == null || userIds.length === 0) {
      return result;
    }
    const userRoles = await this.userRoleRepo.findByUserIds(userIds);
    if (userRoles.length === 0) {
      return result;
    }
    const roleIds = [...new Set(userRoles.map((ur) => ur.roleId))];
    const roles = await this.roleRepo.findByIds(roleIds);
    const roleMap = new Map(roles.map((r) => [r.id!, r]));
    for (const ur of userRoles) {
      const role = roleMap.get(ur.roleId);
      if (role) {
        const list = result.get(ur.userId) ?? [];
        list.push(role);
        result.set(ur.userId, list);
      }
    }
    return result;
  }

  async assertCanDisableUser(targetUserId: number, currentUserId: number): Promise<void> {
    if (targetUserId === currentUserId) {
      throw new BusinessException(ErrorCode.CANNOT_DISABLE_SELF);
    }
    await this.assertNotLastAdmin(targetUserId);
  }

  /** 在事务内锁 ADMIN 绑定后检查目标是否为最后一名活跃管理员。 */
  private async assertNotLastAdmin(targetUserId: number): Promise<void> {
    const adminRole = await this.getAdminRole();
    if (!adminRole || !(await this.userHasRole(targetUserId, adminRole.id!))) {
      return;
    }
    if (!this.userRoleRepo.transaction || !this.userRoleRepo.findByRoleIdForUpdate) {
      if ((await this.countOtherActiveAdmins(adminRole.id!, targetUserId)) === 0) {
        throw new BusinessException(ErrorCode.CANNOT_REMOVE_LAST_ADMIN);
      }
      return;
    }
    await this.userRoleRepo.transaction(async (tx) => {
      if (!tx.findByRoleIdForUpdate) return;
      const bindings = await tx.findByRoleIdForUpdate(adminRole.id!);
      let otherActive = 0;
      for (const b of bindings) {
        if (b.userId === targetUserId) continue;
        const u = await this.userRepo.findById(b.userId);
        if (u && u.status === 1) otherActive += 1;
      }
      if (otherActive === 0) {
        throw new BusinessException(ErrorCode.CANNOT_REMOVE_LAST_ADMIN);
      }
    });
  }

  async assertCanChangeRoles(userId: number, newRoleIds: number[]): Promise<void> {
    const adminRole = await this.getAdminRole();
    if (!adminRole) {
      return;
    }
    const hadAdmin = await this.userHasRole(userId, adminRole.id!);
    const willHaveAdmin = newRoleIds.includes(adminRole.id!);
    if (hadAdmin && !willHaveAdmin && (await this.countOtherActiveAdmins(adminRole.id!, userId)) === 0) {
      throw new BusinessException(ErrorCode.CANNOT_REMOVE_LAST_ADMIN);
    }
  }

  private getAdminRole(): Promise<Role | null> {
    return this.roleRepo.findByCode('ADMIN');
  }

  async userHasRole(userId: number, roleId: number): Promise<boolean> {
    return (await this.userRoleRepo.countByUserAndRole(userId, roleId)) > 0;
  }

  private async countOtherActiveAdmins(adminRoleId: number, excludeUserId: number): Promise<number> {
    const bindings = await this.userRoleRepo.findByRoleId(adminRoleId);
    let count = 0;
    for (const b of bindings) {
      if (b.userId === excludeUserId) {
        continue;
      }
      const user = await this.userRepo.findById(b.userId);
      if (user && user.status === 1) {
        count += 1;
      }
    }
    return count;
  }

  async isAdmin(userId: number | null | undefined): Promise<boolean> {
    if (userId == null) {
      return false;
    }
    const adminRole = await this.getAdminRole();
    return adminRole != null && (await this.userHasRole(userId, adminRole.id!));
  }

  async hasPermission(userId: number, permissionCode: string): Promise<boolean> {
    const userRoles = await this.userRoleRepo.findByUserId(userId);
    if (userRoles.length === 0) {
      return false;
    }
    const roleIds = userRoles.map((r) => r.roleId);
    const rolePerms = await this.rolePermissionRepo.findByRoleIds(roleIds);
    if (rolePerms.length === 0) {
      return false;
    }
    const permIds = rolePerms.map((p) => p.permissionId);
    const count = await this.permissionRepo.countByIdsAndCode(permIds, permissionCode);
    return count > 0;
  }

  async getUserPermissionCodes(userId: number): Promise<string[]> {
    const userRoles = await this.userRoleRepo.findByUserId(userId);
    if (userRoles.length === 0) {
      return [];
    }
    const roleIds = userRoles.map((r) => r.roleId);
    const rolePerms = await this.rolePermissionRepo.findByRoleIds(roleIds);
    if (rolePerms.length === 0) {
      return [];
    }
    const permIds = rolePerms.map((p) => p.permissionId);
    const permissions = await this.permissionRepo.findByIds(permIds);
    return permissions.map((p) => p.code);
  }
}
